File: src/engine/proof.ts
import { SITE } from '@/config/site';
import { confidenceLabel, jarFraction, type BoundDef } from '@/math/bounds';
import type { ModelRun } from './types';

export type GateKey = 'n_samples' | 'n_positive' | 'auc_std' | 'time_split';

type ModelStats = Pick<ModelRun, 'n' | 'nPositive' | 'auc' | 'aucStd' | 'd' | 'bootLower' | 'timeSplitGap' | 'gatesOverride'>;

// #region stage:jar
export function evaluateModel(m: ModelStats, bound: BoundDef) {
  const { epsilon, floor, pending } = bound.compute({
    n: m.n,
    nPos: m.nPositive,
    auc: m.auc,
    aucStd: m.aucStd,
    d: m.d,
    delta: SITE.delta,
    bootLower: m.bootLower,
  });

  // Every gate must pass before the jar may read 100%.
  const g = SITE.gates;
  const gates: Record<GateKey, boolean> = {
    n_samples: m.n >= g.nSamplesMin,
    n_positive: m.nPositive >= g.nPositiveMin,
    auc_std: m.aucStd < g.aucStdMax,
    time_split: m.timeSplitGap <= g.timeSplitGapMax,
    ...(m.gatesOverride as Partial<Record<GateKey, boolean>> | undefined),
  };
  const blockedBy = (Object.keys(gates) as GateKey[]).find((k) => !gates[k]) ?? null;

  const rawJar = jarFraction(floor, SITE.aucFloor, SITE.aucTarget);
  const jar = blockedBy ? Math.min(rawJar, g.jarCapWhenBlocked) : rawJar;
  const unlocked = jar >= 1; // floor ≥ 0.60 and nothing blocking: launch allowed

  return { epsilon, floor, pending: !!pending, gates, blockedBy, rawJar, jar, unlocked, confidence: confidenceLabel(floor) };
}
// #endregion

export type ProofView = ReturnType<typeof evaluateModel>;

File: backend/wassily/engine/proof.py
"""The jar: a bound's floor, capped until every validation gate passes."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol

from wassily.config.site import SITE, Site
from wassily.maths.bounds import BoundDef, BoundInput, confidence_label, jar_fraction

GATE_KEYS = ("n_samples", "n_positive", "auc_std", "time_split")


class ModelStats(Protocol):
    n: int
    n_positive: int
    auc: float
    auc_std: float
    d: int
    boot_lower: float
    time_split_gap: float
    gates_override: dict[str, bool] | None


@dataclass(frozen=True)
class ProofView:
    epsilon: float
    floor: float
    pending: bool
    gates: dict[str, bool]
    blocked_by: str | None
    raw_jar: float
    jar: float
    unlocked: bool  # floor >= target and nothing blocking: a launch is allowed
    confidence: str


def evaluate_model(m: ModelStats, bound: BoundDef, site: Site = SITE) -> ProofView:
    result = bound.compute(
        BoundInput(n=m.n, n_pos=m.n_positive, auc=m.auc, auc_std=m.auc_std, d=m.d, delta=site.delta,
                   boot_lower=m.boot_lower)
    )

    # Every gate must pass before the jar may read 100%.
    g = site.gates
    gates = {
        "n_samples": m.n >= g.n_samples_min,
        "n_positive": m.n_positive >= g.n_positive_min,
        "auc_std": m.auc_std < g.auc_std_max,
        "time_split": m.time_split_gap <= g.time_split_gap_max,
    }
    if getattr(m, "gates_override", None):
        gates.update(m.gates_override or {})
    blocked_by = next((k for k in gates if not gates[k]), None)

    raw_jar = jar_fraction(result.floor, site.auc_floor, site.auc_target)
    jar = min(raw_jar, g.jar_cap_when_blocked) if blocked_by else raw_jar

    return ProofView(
        epsilon=result.epsilon,
        floor=result.floor,
        pending=result.pending,
        gates=gates,
        blocked_by=blocked_by,
        raw_jar=raw_jar,
        jar=jar,
        unlocked=jar >= 1,
        confidence=confidence_label(result.floor),
    )
